File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode reverse(ListNode node){
        ListNode prev=null;
        ListNode cur=node;
        ListNode next=null;
        while(cur!=null){
            
            next=cur.next;
            cur.next=prev;
            prev=cur;
            cur=next;
        }
        return prev;
    }
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        ListNode ans=dfs(l1,l2,0);
        return ans;
    }
    ListNode dfs(ListNode l1,ListNode l2,int carry){
       if(l1==null && l2==null && carry==0){
        return null;
       }
       int temp1=l1!=null?l1.val:0;
       int temp2=l2!=null?l2.val:0;
       ListNode l1next=l1==null?null:l1.next;
       ListNode l2next=l2==null?null:l2.next;
       int ans=temp1+temp2+carry;
       
       ListNode cur=new ListNode(ans%10);
       ListNode next=dfs(l1next,l2next,ans/10);
       cur.next=next;
       return cur;
    }
}

File: 2-add-two-numbers/add-two-numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    int carry = 0;
    struct ListNode *l3 = NULL, *tail = NULL;

    while (l1 || l2 || carry) {
        int sum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
        int digit = sum % 10;
        carry = sum / 10;

        struct ListNode *newNode = (struct ListNode *)malloc(sizeof(struct ListNode));
        newNode->val = digit;
        newNode->next = NULL;

        if (tail) {
            tail->next = newNode;
            tail = newNode;
        } else {
            l3 = tail = newNode;
        }

        if (l1) l1 = l1->next;
        if (l2) l2 = l2->next;
    }

    return l3;
}